persistence/src/main/java/de/cyface/persistence/Persistence.java

```diff
@@ -0,0 +1,682 @@
+package de.cyface.persistence;
+
+import static de.cyface.persistence.Constants.TAG;
+import static de.cyface.persistence.serialization.MeasurementSerializer.TRANSFER_FILE_FORMAT_VERSION;
+import static de.cyface.persistence.serialization.MeasurementSerializer.serialize;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.zip.Deflater;
+
+import android.content.ContentResolver;
+import android.content.ContentValues;
+import android.content.Context;
+import android.database.Cursor;
+import android.net.Uri;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import de.cyface.persistence.model.GeoLocation;
+import de.cyface.persistence.model.Measurement;
+import de.cyface.persistence.model.Vehicle;
+import de.cyface.persistence.serialization.*;
+import de.cyface.utils.Validate;
+
+/**
+ * This class wraps the Cyface Android persistence API as required by the <code>DataCapturingService</code>,
+ * <code>SyncAdapter</code> and its
+ * delegate objects.
+ *
+ * @author Armin Schnabel
+ * @version 1.0.2
+ * @since 3.0.0
+ */
+public class Persistence {
```

Comment:
I really dislike the idea that you can create instances of this and also instances of the subclass `MeasurementPersistence`. Creating instances of non-leaf classes (with respect to the inheritance tree) is really confusing.

Reply:
If I understand you correctly:

The Constructor of the **non-leaf Persistence class** is used in the SyncAdapter.onPerformSync() in order to load the measurements from the persistence layer, to serialize them and to mark them as synced.

1) If we would use the **leaf class MeasurementPersistence** then a threadpool (used for capturing) would be created which we don't need, which is the only real difference between the constructors.

2) The other difference is that MeasurementPersistence caches a reference to the currentlyCapturedMeasurement which should not really be a problem for the synchronization as the synchronization does not create a new Measurement (which is where the cache is set) so it would just be null all the time which is ok I guess?

**Any suggestions how we could handle this differently in a clean way?**

The name of this class is discussed in a comment above.

Reply:
The thing is that clean code requires classes to either be designed for inheritance. In that case it should not be possible to create instances and required to create a child with a concrete implementation. So if there really is a difference we should do something like:

Persistence
|-> NormalPersistence
|-> MeasurementPersistence

If the difference is very superficial we should think about our architecture. Do we really need this differentiation? Is there a place where you provide a reference to a Persistence which can be filled by either NormalPersistence or MeasurementPersistence? If there is not, we should prefer delegation over inheritance. So in that case we move everything that is different between MeasurementPersistence and NormalPersistence to two new classes inheriting from a common ancestor and provide a reference to that ancestor to the constructor of Persistence (MeasurementPersistence and NormalPersistence would not exist in that case). I hope this is kind of clear. In the end we would have:

PersistenceLayer (no subclasses)
- <<attribute>> behaviour: PersistenceBehaviour
+ <<constructor>> PersistenceLayer(PersistenceBehaviour)

PersistenceBehaviour (please think of a proper name)
|-> NormalPersistenceBehaviour
| -> MeasurementPersistenceBehaviour

But you do have better insight if such a setup is even necessary.

Reply:
The behavioral pattern is such a beauty here ...